### DISCUSSION TITLE
getNodes 不满意使用场景，有类似 getTreeNodes 的实现 API 吗 ？ 另外 React 在 X6 中如何进行数据管理？

### DISCUSSION THREAD

**Discussion by mist0616** [antvis/X6#3521] [Q&A]
第一次使用 X6，目前在做一个以表单为节点的思维导图，表单数据通过 node 的 getData、setData 管理，保存时在最外层使用 getNodes + getData 获取，但因为 getNodes 是平铺的，无法满足需求，想问一下如何获取带层级的节点信息？顺便问下官方对于 React 在 X6 中的数据管理有最佳实践教学吗？

**Comment by ahuigo** [comment]
1. getTreeNodes　这个可能需要自己按自己的业务需要自行封装
2. react 数据管理，一般是：
　1.　利用context：参考：https://x6.antv.antgroup.com/tutorial/intermediate/react　中使用potal 将数据传到x6 内部去。
示例：https://github.com/ahuigo/umi-demo/blob/main/src/pages/x6/node/node-react-portal.tsx
　2.　利用redux（复杂）
　3.　利用useSyncExternalStore (简单)：https://github.com/ahuigo/umi-demo/blob/main/src/pages/react/context-resso.tsx
　4. 　以及其它的工具

**Comment by lloydzhou** [comment]
@misty-zq 

推荐一个用于管理x6数据的组件
https://github.com/lloydzhou/x6-hooks

提供一个`useGraphState`hooks，使用react hooks的接口操作节点和边数据。这个hooks内部有一个`diff`/`patch`的逻辑把数据自动更新到x6画布上面
